File: rules/lib/rule.ts
import { z } from 'zod';
import {
  OsConstraintSchema,
  type OsConstraint,
  type OsOptions,
  satisfiesOs,
} from './os';
import {
  FeatureConstraintSchema,
  type FeatureConstraint,
  satisfiesFeatures,
} from './features';

export type RuleAction = 'allow' | 'disallow';

export type Rule =
  | { action: RuleAction; os: OsConstraint }
  | { action: RuleAction; features: FeatureConstraint }
  | { action: RuleAction };

export const RuleSchema: z.ZodType<Rule> = z.union([
  z.object({ action: z.string(), os: OsConstraintSchema }),
  z.object({ action: z.string(), features: FeatureConstraintSchema }),
  z.object({ action: z.string() }),
]) as z.ZodType<Rule>;

export function satisfiesRule(
  rule: Rule,
  os: OsOptions,
  feats: string[] = [],
): boolean {
  const allow = rule.action === 'allow';
  if ('os' in rule) return satisfiesOs(rule.os, os) === allow;
  if ('features' in rule)
    return satisfiesFeatures(rule.features, feats) === allow;
  return allow;
}

File: crates/opys-mojang-rules/src/rule.rs
use serde::{Deserialize, Serialize};

use crate::features::{satisfies_features, FeatureConstraint};
use crate::os::{satisfies_os, OsConstraint, OsOptions, RuleError};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RuleAction {
    Allow,
    Disallow,
}

/// A rule is `{ action }` plus an optional `os` or `features` constraint.
///
/// Order in `untagged` matters: the OS variant checks first, then features,
/// then plain. JSON with extra keys is tolerated (matches the TS `z.object`
/// strip-unknown behavior).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Rule {
    Os {
        action: RuleAction,
        os: OsConstraint,
    },
    Features {
        action: RuleAction,
        features: FeatureConstraint,
    },
    Plain {
        action: RuleAction,
    },
}

impl Rule {
    pub fn action(&self) -> RuleAction {
        match self {
            Rule::Os { action, .. } | Rule::Features { action, .. } | Rule::Plain { action } => {
                *action
            }
        }
    }
}

pub fn satisfies_rule(
    rule: &Rule,
    os: &OsOptions,
    feats: &[String],
) -> Result<bool, RuleError> {
    let allow = matches!(rule.action(), RuleAction::Allow);
    match rule {
        Rule::Os { os: c, .. } => Ok(satisfies_os(c, os)? == allow),
        Rule::Features { features, .. } => Ok(satisfies_features(features, feats) == allow),
        Rule::Plain { .. } => Ok(allow),
    }
}
